easyCore/Datasets/xarray.py

```diff
@@ -0,0 +1,765 @@
+        :type new_computational_fn: Callable
+        :return: None
+        :rtype: None
+        """
+        self._obj.attrs['computation']['postcompute_func'] = new_computational_fn
+
+    def fit_prep(self, func_in: Callable, bdims=None, dask_chunks=None) -> Tuple[xr.DataArray, Callable]:
+        """
+        Generate boradcasted coordinates for fitting and reform the fitting function into one which can handle xarrays
+
+        :param func_in: Function to be wrapped and made xarray fitting compatable.
+        :type func_in: Callable
+        :param bdims: Optional precomputed broadcasted dimensions.
+        :type bdims: xr.DataArray
+        :param dask_chunks: How to split to broadcasted dimensions for dask.
+        :type dask_chunks: Tuple[int..]
+        :return: Tuple of broadcasted fit arrays and wrapped fit function.
+        :rtype: xr.DataArray, Callable
+        """
+
+        if bdims is None:
+            coords = [self._obj.coords[da].transpose() for da in self._obj.dims]
+            bdims = xr.broadcast(*coords)
+        self._obj.attrs['computation']['compute_func'] = func_in
+
+        def func(x, *args, vectorize: bool = False, **kwargs):
+            old_shape = x.shape
+            if not vectorize:
+                xs = [x_new.flatten() for x_new in [x, *args] if isinstance(x_new, np.ndarray)]
+                x_new = np.column_stack(xs)
+                if len(x_new.shape) > 1 and x_new.shape[1] == 1:
+                    x_new = x_new.reshape((-1))
+                result = self.compute_func(x_new, **kwargs)
+            else:
+                result = self.compute_func(*[d for d in [x, args] if isinstance(d, np.ndarray)],
+                                 *[d for d in args if not isinstance(d, np.ndarray)],
+                                 **kwargs)
+            if isinstance(result, np.ndarray):
+                result = result.reshape(old_shape)
+            result = self.postcompute_func(result)
+            return result
+
+        return bdims, func
+
+    def generate_points(self) -> xr.DataArray:
+        """
+        Generate an expanded DataArray of points which corresponds to broadcasted dimensions (`all_x`) which have been
+        concatenated along the second axis (`fit_dim`).
+
+        :return: Broadcasted and concatenated coordinates
+        :rtype: xr.DataArray
+        """
+
+        coords = [self._obj.coords[da] for da in self._obj.dims]
+        c_array = []
+        n_array = []
+        for da in xr.broadcast(*coords):
+            c_array.append(da)
+            n_array.append(da.name)
+
+        f = xr.concat(c_array, dim='fit_dim')
+        f = f.stack(all_x=n_array)
+        return f
+
+    def fit(self, fitter, *args,
+            fit_kwargs: dict = None,
+            fn_kwargs: dict = None,
+            vectorize: bool = False,
+            dask: str = 'forbidden',
+            **kwargs) -> FitResults:
+        """
```

Comment:
I think the argument list of `fit`, both here and down in Array, could use some love.
Fit funtion passed as `fitter` takes `*args` immediately after it, but its `**kwargs` are hiding in keyworded `fn_kwargs`.
`**kwargs` actually are for `xarray.apply_ufunc` and we also have `fit_kwargs` for control.

I am not sure how to streamline it nicely but maybe adding some containers for each set of arguments would make sense?
E.g.
```python
def fit(self,
          fitter,
          fitter_args=[*args, **kwargs],
          ufunc_args: dict=None,
          control_args:dict=None
         ...
```

I especially find the final `**kwargs` confusing - I would expect these to be passed to the fitter...